### ISSUE TITLE
Multiplexed data records

### ISSUE THREAD

**Issue by chad-iris** [iris-edu/xseed-specification#21]
A discussion on design allowing multiplexed, i.e. more than one channel within, a record.

### preamble
A key design of this format is simplicity, which is extremely valuable for a long-term format.  The current design fits the major use cases of archive and exchange quite well (as simple as possible but not simpler).  The WG II requirements identify real-time data transfer as "not a major goal, but NGF should be designed to reasonabl[y] support real-time transfer without compromising other use-cases".

It would be valuable to explore the potential for multiplexed data, in such as way that minimizes added complexity to the design, to enhance usefulness for real-time transfer.

A key aspect of such multiplexed data would be the ability to re-combine it into non-multiplexed data for the archive and exchange use-cases.

### scope of multiplexing
There are direct trade-offs between how flexible the multiplexing, how complex the scheme, and how efficient/redundant the contained information.  On one extreme, as employed by some dataloggers, all channels are contained in the same "record".  This requires storing sample rates, counts, encodings?, CRCs?, etc. for each payload.  On the other extreme would be only multiplexing channels that have the same sample rates, counts, encodings, etc. to take advantage of their commonalities.  In the middle would be an option to group payloads with commonalities, but this requires more complexity, i.e. more data structures.

Based on discussions at the Montreal meeting and a desire minimize complexity while allowing (ideally trivial) re-combination to a non-multiplexed for for archiving, I'm leaning towards less flexible multiplexing.

### an encoding for multiplexed payloads and prefix-suffix identifiers
The main concepts:
* Define encodings for a multiplexed payloads.  Perhaps as simple as adding 100 to each defined payload such that any encoding >= 100 is a multiplexed version of the normal encoding.  For example, a value of `103` is multiplexed payloads of 32-bit integers, which have an encoding of value `3`.
* Split the identifiers into a _prefix_ (contained in the main header) and a _suffix_ (contained in the payload.

The same fixed header is used and all fields have the same meaning, with the knowledge that the source identifier is not completely know until the payloads are traversed.  The payloads all share the same sample rate, count, and encoding.  The length of the data payload field is the total length of all (sub)payloads.

Each (sub)payload would need some minimal structure, such as:

|Field |Name|Type|Length|Offset|
|---|---|---|---|---|
|1|Length of data payload in bytes    |UINT16  |2  | 0  |
|2|Length of identifier suffix in bytes|UINT8    |1  | 2  |
|3|Source identifier suffix                   |CHAR     |V  | 3  |
|4|Data payload                                   |encoded|V  | 3 + field2|

For example, a record header could have an identifier of `XFDSN:IU_COLA_00_BH`, with 3 (sub)payloads with identifier suffixes of `Z`, `1`, and `2`.

Importantly, this kind of record could be re-combined into a non-multiplexed version pretty easily.  The fundamental type encodings (0-5) allow trivial combination of data (sub)payloads.  Unfortunately, Steim encodings would need to be decoded to be combined I believe; this might be a good characteristic to keep in mind for #4.

**Comment by crotwell** [comment]
Obviously adding multiplexing adds significant complexity to the format, and so there should be a significant benefit to justify it. But without a clearer problem description, it is hard for me to evaluate whether the solution is worth the cost.

I assume near realtime transmission is it, but can you flesh that out in terms of why multiplexing is needed? Is it just to reduce the overhead of the fixed header when the number of samples is small? Or are there other benefits or other use cases where multiplexing n channels data together would be a significant advantage? 

There are two main alternatives to multiplexing within a single record that I see. Trivially of course is just to demultiplex and create n records sequentially. The second alternative, if realtime transmission is the only use case, is to create a separate protocol whose sole purpose is efficient transmission of realtime data, and is not concerned with archiving. This protocol might allow significant benefits over a multiplexed record while not complicating the record format for archiving purposes. It could share many features with the xseed record structure where it makes sense, but not be constrained by the archival needs.

As an very simplistic example, perhaps the protocol allows sending of entire xseed records periodically or as needed due to timing issues and then in between it sends an "append data to existing record" packet which has minimal overhead. The sequence might go like:

1. send xseed for BHZ with 1 sample, label as channel A
2. send xseed for BHE with 1 sample, label as channel B
3. send xseed for BHN with 1 sample, label as channel C
4. append [2,5,] to  A
5. append [-1, 3] to C
6. append [4,8] to B
7. ...repeat appends 
8. send xseed for BHE with 1 sample, label as channel B (ending previous BHE record)
9. append [-2, 4,2,8] to C
10. append [-3,-5] to B
11. ...

Would such a protocol be worth thinking about as an alternative?

**Comment by chad-iris** [comment]
> I assume near realtime transmission is it, but can you flesh that out in terms of why multiplexing is needed? Is it just to reduce the overhead of the fixed header when the number of samples is small? Or are there other benefits or other use cases where multiplexing n channels data together would be a significant advantage?

I believe it is primarily to reduce overhead of the fixed header for "naturally grouped" channels, while allowing short payload lengths (e.g. 1 second).  I was asked about multiplexing potential independently by folks.  What I'm unclear on is whether this is _enough_ grouping or if multiplexing of different sample rates, payload sample counts, etc. is desired.

> There are two main alternatives to multiplexing within a single record that I see. Trivially of course is just to demultiplex and create n records sequentially.

Indeed, if one is willing to burn bandwidth on the overhead of headers one could always just create short records.  While short records are not wonderful for the archive, they are way better than multiplexed records.

> The second alternative, if realtime transmission is the only use case, is to create a separate protocol whose sole purpose is efficient transmission of realtime data, and is not concerned with archiving.

The effort here is to see if there is a "good enough" solution for many of the real-time stream use cases that is as easy as possible to convert into the archive/exchange format.  Going through this [thought process again](https://github.com/FDSN/miniSEED3-TechnicalEvaluation/issues/25), is really a last ditch effort at a middle solution because this issue was raised again.  It very well may be better to have a separate format focussed on real-time transmission.

**Comment by crotwell** [comment]
Reviewing that old thread and thinking more, I feel near real time transmission where you care about low latency but not low bandwidth is addressed by the flexible record size, ie non-power of two. The receiver can also consolidate many small records into fewer larger records if desired for archiving in the same way that, for example, ew2ringserver, does currently.

On the other hand, if you want low latency AND low bandwidth, then you really need a system where the sender (data logger) and receiver are tightly coupled, and are free to innovate over much shorter timescales than an FDSN standard can provide, and hence you want a custom/proprietary protocol. 

So, if this comes up for a vote, I vote NO on multiplexing.

**Comment by jmfee-usgs** [comment]
My understanding of the use case is for multiple channels that are acquired in parallel, and are ready for transmission at the same time; same sample rate, same payload size, etc.  This would group data for a sensor with multiple components that are equivalent in all ways except orientation, at least during transmission.

Would adding (optional) compression to the SEEDLink connection provide similar benefits, i.e. reducing duplication within headers, without requiring changes to the format?

**Comment by djeastonca** [comment]
> I was asked about multiplexing potential independently by folks. What I'm unclear on is whether this is enough grouping or if multiplexing of different sample rates, payload sample counts, etc. is desired.

Another use case/motivation for the multiplexing of data that would be ideal to address at some point is to reduce the latency associated with the transmission of time series state of health data from stations without the expense of excessive overhead for the live monitoring of seismic networks. In this case there are more channels involved but the problem is essentially the same, except that the latency requirements are not as tight. In both cases the goal is reducing overhead while minimizing latency. Having said that, the solution for each would not be the same due to the channel data rate differences. State of health channel data rates would typically be once per minute so multiplexing data in this context would entail mixing individual samples that, together, comprise a ‘snapshot’ of instrument state of health at a particular time. This would be the notion of all of the channels contained in the same “record” that @chad-iris referred to.

With regard to multiplexing seismic data, for what it’s worth here is a variation on @crotwell 's earlier approach that skews a little bit towards the flexibility direction of the spectrum of multiplexing solutions while still keeping things simple. When a consumer connects to the producer, the producer would send ‘metadata’ for each of the requested channels that reflects the current configuration of the digitizer/datalogger (channel naming, sample rate, sample encoding, etc). The metadata could be roughly what’s in the data record header except for the timestamp [1]. A sequence number would be needed to distinguish between successive metadata updates in the event of a configuration change in the datalogger. After that would be a sequence of separate encoded data payloads for each channel. Each data payload has a start time and a sequence number and a means of referencing its corresponding metadata. The flexibility this offers is the relax the constraint that all channels share the same metadata and avoids the notion of a ‘base’ channel name with suffixes specified in the payloads:

1. Consumer connects to producer and subscribes to channels (e.g. HHZ, HHN, HHE)
2. Producer sends metadata for subscribed channels HHZ, HHN, HHE
3. Producer sends data packet for HHZ
4. Producer sends data packet for HHN
5. Producer sends data packet for HHE
6. … and so on

A consumer could re-combine the stream into non-multiplexed data for storage/archiving easily since the metadata is readily at hand.

[1] Ideally the channel metadata would be enough for the consumer to construct the full channel response but the problem of communicating full station response information via streaming is out of scope for this discussion.

**Comment by djeastonca** [comment]
> Would adding (optional) compression to the SEEDLink connection provide similar benefits, i.e. reducing duplication within headers, without requiring changes to the format?

Compressing the payloads in the streamed data would be problematic with regard to the sharing of timestamps since the compression rates for the multiplexed channels would be different.

**Comment by chad-iris** [comment]
Thanks for the thoughts.

> When a consumer connects to the producer, the producer would send ‘metadata’ for each of the requested channels that reflects the current configuration of the digitizer/datalogger (channel naming, sample rate, sample encoding, etc). The metadata could be roughly what’s in the data record header except for the timestamp [1].

A scheme like this seams to be what is required to meaningfully reduce overhead while minimizing latency.  Any such scheme needs a notion of state for context, and is well beyond the complexity of miniSEED for an archive.  The data blocks could be reused verbatim in a non-multiplexed format if the header contained the metadata to interpret them.  But this would still be a format conversion and not a trivial re-combination.

**Comment by jmfee-usgs** [comment]
@djeastonca I was thinking compressing the entire stream, using deflate or similar, rather than compressing individual payloads.
